work/worker.go

```diff
@@ -0,0 +1,151 @@
+		WorkerID: 1,
+		Workers:  make(map[int]Worker),
+	}
+}
+
+// WorkerManager controls Workers allowing them to be started and stopped independently of each other.
+type WorkerManager struct {
+	WorkerID  int
+	Workers   map[int]Worker
+	WorkersMu sync.Mutex
+}
+
+type Worker interface {
+	Start(ctx context.Context)
+	Stop(ctx context.Context)
+	Err() error
+	Running() bool
+}
+
+// AddTaskAsWorker wraps schedule.Task `t` as a TaskWorker, adds it to WorkerManager's Worker list,
+// and returns its ID. Workers can be started and stopped using their ID's.
+func (wm *WorkerManager) AddTaskAsWorker(t schedule.Task) int {
+	tw := &TaskWorker{
+		task:    t,
+		running: false,
+	}
+	return wm.AddWorker(tw)
+}
+
+// AddWorker adds Worker `w` to WorkerManager's Worker list and returns its ID. Workers can be
+// started and stopped using their ID's.
+func (wm *WorkerManager) AddWorker(w Worker) int {
+	wm.WorkersMu.Lock()
+	defer func() {
+		wm.Workers[wm.WorkerID] = w
+		wm.WorkerID++
+		wm.WorkersMu.Unlock()
+	}()
+	log.Infow("added worker", "id", wm.WorkerID)
+	return wm.WorkerID
+}
+
+// Start starts the worker with ID `id`, errors if the ID is not found, or id the worker us running.
+func (wm *WorkerManager) Start(ctx context.Context, id int) error {
+	wm.WorkersMu.Lock()
+	defer wm.WorkersMu.Unlock()
+
+	w, ok := wm.Workers[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
+	if w.Running() {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	log.Infow("starting worker", "id", id)
+	w.Start(ctx)
+	return nil
+}
+
+// Stop stops the worker with ID `id`, errors if the ID is not found or if the worker is not running.
+func (wm *WorkerManager) Stop(ctx context.Context, id int) error {
+	wm.WorkersMu.Lock()
+	defer wm.WorkersMu.Unlock()
+
+	w, ok := wm.Workers[id]
+	if !ok {
+		return xerrors.Errorf("stopping worker ID: %d not found", id)
+	}
+	if !w.Running() {
+		return xerrors.Errorf("stopping worker ID: %d is not running", id)
+	}
+	log.Infow("stopping worker", "id", id)
+	w.Stop(ctx)
+	return nil
+}
+
+// TaskWorker wraps a schedule.Task to implement the Worker interface
+type TaskWorker struct {
+	task    schedule.Task
+	err     error
+	mu      sync.Mutex
+	running bool
+	cancel  func()
+}
+
+// Running returns true if the Worker is active, false otherwise.
+func (tw *TaskWorker) Running() bool {
+	return tw.running
+}
+
+// Start is non-block and runs the Worker until Stop is called, the context is done, or an error occurs.
+func (tw *TaskWorker) Start(ctx context.Context) {
+	cctx, cancel := context.WithCancel(ctx)
+	tw.initialize(cancel)
+	go func(c context.Context) {
+		if err := tw.task.Run(c); err != nil {
+			tw.shutdown(err)
+		}
+	}(cctx)
+}
+
+func (tw *TaskWorker) initialize(cancel context.CancelFunc) {
+	tw.mu.Lock()
+	tw.running = true
+	tw.err = nil
+	tw.cancel = cancel
+	tw.mu.Unlock()
+}
+
+func (tw *TaskWorker) shutdown(err error) {
+	tw.mu.Lock()
+	tw.cancel()
```

Comment:
If `Stop` is called before `Start` then this will panic. Check `tw.running == true` first.